### PR TITLE
[Avro] Use "extensions/avro" instead of avro from"core" in Java SDK modules

### LINKED ISSUE #24748: [Task]: [Avro] Use "extensions/avro" instead of avro from"core" in Java SDK modules
### What needs to happen?

Once Avro support is extracted into a dedicated module `extensions/avro`, then all Beam Java SDK modules have to switch to use it instead of avro-related classes from "core".

Affected packages:
- org.apache.beam.examples
- org.apache.beam.runners.dataflow
- org.apache.beam.sdk.extensions.sql
- org.apache.beam.sdk.io.gcp
- org.apache.beam.sdk.io.kafka
- org.apache.beam.sdk.io.parquet
- org.apache.beam.sdk.nexmark

### Issue Priority

Priority: 2 (default / most normal work should be filed as P2)

### Issue Components

- [ ] Component: Python SDK
- [X] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by aromanenko-dev**
Make all Beam Java SDK modules, that depend on Avro, use Avro extension instead of Avro from `core`.

It should not introduce any breaking changes for users.

Closes #24748

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by olehborysevych** [comment]
@aromanenko-dev looking into failing Example check. Sorry for inconvenience (

**Comment by aromanenko-dev** [comment]
Run Java_Kafka_IO_Direct PreCommit

**Comment by aromanenko-dev** [comment]
Run SQL_Java17 PreCommit

**Comment by aromanenko-dev** [comment]
Run Java_GCP_IO_Direct PreCommit

**Comment by aromanenko-dev** [comment]
Run SQL PreCommit

**Comment by aromanenko-dev** [comment]
Run Java_Kafka_IO_Direct PreCommit

**Comment by aromanenko-dev** [comment]
Run SQL PreCommit

**Comment by aromanenko-dev** [comment]
Run Java_Pulsar_IO_Direct PreCommit

**Comment by aromanenko-dev** [comment]
Run Java_GCP_IO_Direct PreCommit

**Comment by aromanenko-dev** [comment]
Run Java PreCommit

**Comment by aromanenko-dev** [comment]
Run SQL_Java11 PreCommit

**Comment by aromanenko-dev** [comment]
retest this please

**Comment by aromanenko-dev** [comment]
Run Java PreCommit

**Comment by aromanenko-dev** [comment]
Run Java_Pulsar_IO_Direct PreCommit

**Comment by aromanenko-dev** [comment]
Run Java PostCommit

**Comment by aromanenko-dev** [comment]
Run Java PreCommit

**Comment by aromanenko-dev** [comment]
Run Java PreCommit

**Comment by aromanenko-dev** [comment]
Run Java PreCommit

**Comment by aromanenko-dev** [comment]
Run Java PreCommit

**Comment by aromanenko-dev** [comment]
Run Java_GCP_IO_Direct PreCommit

**Comment by aromanenko-dev** [comment]
Run Java PostCommit

**Comment by aromanenko-dev** [comment]
Run Spark ValidatesRunner

**Comment by aromanenko-dev** [comment]
Run Direct ValidatesRunner

**Comment by aromanenko-dev** [comment]
Run Dataflow ValidatesRunner

**Comment by aromanenko-dev** [comment]
Run Flink ValidatesRunner

**Comment by aromanenko-dev** [comment]
Run Java PostCommit

**Comment by aromanenko-dev** [comment]
Run Spark ValidatesRunner

**Comment by aromanenko-dev** [comment]
Run Dataflow ValidatesRunner

**Comment by aromanenko-dev** [comment]
Run Flink ValidatesRunner

**Comment by aromanenko-dev** [comment]
Run Spark Runner Nexmark Tests

**Comment by aromanenko-dev** [comment]
Run Dataflow Runner V2 Nexmark Tests

**Comment by aromanenko-dev** [comment]
Run Spark StructuredStreaming ValidatesRunner

**Comment by aromanenko-dev** [comment]
R: @mosche @reuvenlax @lukecwik 
CC: @kennknowles 

Please, take a look and let me know if you have any comments/objections and recommendations for testing that should be done before accepting this PR.

**Comment by mosche** [comment]
@aromanenko-dev Don't forget to enable publishing of the extension
```
applyJavaNature(
    automaticModuleName: 'org.apache.beam.sdk.extensions.avro',
    disableLintWarnings: ['rawtypes'], // Avro-generated test code has raw-type errors
    publish: false, 
    exportJavadoc: false,
)
```

File: sdks/java/io/kafka/build.gradle [hunk 1]
```diff
@@ -90,6 +91,7 @@ dependencies {
   provided library.java.everit_json_schema
   testImplementation project(path: ":sdks:java:core", configuration: "shadowTest")
   testImplementation project(":sdks:java:io:synthetic")
+  testImplementation project(path: ":sdks:java:extensions:avro", configuration: "testRuntimeMigration")
```
Comment (mosche) [inline sdks/java/io/kafka/build.gradle:94]: Shouldn't `implementation` be enough?

File: runners/spark/3/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/SparkSessionFactory.java [hunk 2]
```diff
@@ -70,6 +68,8 @@
 import org.apache.beam.sdk.coders.VarIntCoder;
 import org.apache.beam.sdk.coders.VarLongCoder;
 import org.apache.beam.sdk.coders.VoidCoder;
+import org.apache.beam.sdk.extensions.avro.coders.AvroCoder;
```
Comment (mosche) [inline runners/spark/3/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/SparkSessionFactory.java:71]: How about only registering these if they exist on the classpath? So we can skip adding the dependency to the runner by default.

**Review by mosche — COMMENTED** [review]
🎉 Some comments @aromanenko-dev. Wondering, does it make sense to also deprecate the Avro stuff in core in the same PR?

**Comment by aromanenko-dev** [comment]
@mosche 
> Wondering, does it make sense to also deprecate the Avro stuff in core in the same PR?

I wanted to do this in a separate PR once this one is merged. Do you think it'd be better to do here as well?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/CountingSource.java [hunk 3]
```diff
@@ -481,11 +489,45 @@ public long getSplitBacklogBytes() {
     }
   }
 
+  public static class CounterMarkCoder extends CustomCoder<CounterMark> {
```
[the remark this answers is not in the document]
Comment (aromanenko-dev) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/io/CountingSource.java:492]: Anyway, we should avoid stop using Avro in `core`.

File: sdks/java/extensions/avro/src/test/java/org/apache/beam/sdk/extensions/avro/coders/CoderRegistryTest.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Tests for CoderRegistry and AvroCoder. */
+@RunWith(JUnit4.class)
+public class CoderRegistryTest {
+
```
[the remark this answers is not in the document]
Comment (aromanenko-dev) [reply sdks/java/extensions/avro/src/test/java/org/apache/beam/sdk/extensions/avro/coders/CoderRegistryTest.java:50]: There is a small part that checks that `AvroCoder` is properly loaded with `CoderRegistry` when using  `@DefaultCoder(AvroCoder.class)`. I agree that it's not Avro-specific, so can probably just remove it.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/CountingSource.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mosche) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/io/CountingSource.java:492]: Sure, of course, but that's not the point I'm making. Switching coders is sketchy and can cause trouble if the encoded bytes are persisted somehow.
Assuming somebody migrates a pipeline that uses checkpoints containing an Avro encoded `CounterMark` (unlikely, but possible), switching the coder here without being binary compatible will cause issues when migrating to the new Beam version...

**Comment by mosche** [comment]
> I wanted to do this in a separate PR once this one is merged. Do you think it'd be better to do here as well?

@aromanenko-dev I suppose that's fine. Even worst case, if it ends up in another release, there's no harm. The change as is is fairly transparent. 👍 Though it would be good to document usage of the new module in the changelog.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/io/Providers.java [hunk 5]
```diff
@@ -42,12 +42,30 @@ private Providers() {}
   public static <T extends Identifyable> Map<String, T> loadProviders(Class<T> klass) {
     Map<String, T> providers = new HashMap<>();
     for (T provider : ServiceLoader.load(klass)) {
-      checkArgument(
-          !providers.containsKey(provider.identifier()),
-          "Duplicate providers exist with identifier `%s` for class %s.",
-          provider.identifier(),
-          klass);
-      providers.put(provider.identifier(), provider);
+      // Avro provider is treated as a special case until two providers may exist: in "core"
```
[the remark this answers is not in the document]
Comment (aromanenko-dev) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/io/Providers.java]: I updated a comment to make it more clear.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/io/Providers.java [hunk 6]
```diff
@@ -42,12 +42,30 @@ private Providers() {}
...
+      // (deprecated) and in "extensions/avro" (actual).
+      if (provider.identifier().equals("avro")) {
+        // Avro provider from "extensions/avro" must have a priority.
+        if (provider
```
[the remark this answers is not in the document]
Comment (aromanenko-dev) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/io/Providers.java:49]: Good catch, done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/io/Providers.java [hunk 7]
```diff
@@ -42,12 +42,30 @@ private Providers() {}
...
+            .toString()
+            .startsWith(
+                "org.apache.beam.sdk.extensions.avro.schemas.io.payloads.AvroPayloadSerializerProvider")) {
+          // Use AvroPayloadSerializerProvider from extensions/avro by any case.
+          providers.put(provider.identifier(), provider);
+        } else {
+          // Load Avro provider from "core" if it was not loaded from Avro extension before.
+          if (!providers.containsKey(provider.identifier())) {
```
[the remark this answers is not in the document]
Comment (aromanenko-dev) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/io/Providers.java]: Good catch, done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/io/Providers.java [hunk 8]
```diff
@@ -42,12 +42,30 @@ private Providers() {}
...
+            providers.put(provider.identifier(), provider);
+          }
+        }
+      } else {
+        checkArgument(
```
[the remark this answers is not in the document]
Comment (aromanenko-dev) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/io/Providers.java]: Agree, done

**Comment by aromanenko-dev** [comment]
Run Java PreCommit

File: runners/spark/3/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/SparkSessionFactory.java (hunk 2 above)
Comment (aromanenko-dev) [reply runners/spark/3/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/SparkSessionFactory.java]: Makes sense, done

File: sdks/java/io/kafka/build.gradle (hunk 1 above)
Comment (aromanenko-dev) [reply sdks/java/io/kafka/build.gradle]: Yes, done

**Comment by aromanenko-dev** [comment]
Run Java_Pulsar_IO_Direct PreCommit

File: runners/spark/3/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/SparkSessionFactory.java [hunk 9]
```diff
@@ -283,5 +283,13 @@ public void registerClasses(Kryo kryo) {
       kryo.register(TupleTag.class);
       kryo.register(TupleTagList.class);
     }
+
+    private void tryToRegister(Kryo kryo, String className) {
+      try {
+        kryo.register(Class.forName(className));
+      } catch (ClassNotFoundException e) {
+        LOG.warn("Class {}} was not found on classpath", className);
```
Comment (mosche) [inline runners/spark/3/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/SparkSessionFactory.java:291]: Nit, just log these as info? It's kind of expected that these are missing.

File: runners/core-construction-java/build.gradle [hunk 10]
```diff
@@ -73,5 +74,6 @@ dependencies {
   testImplementation library.java.jackson_dataformat_yaml
   testImplementation project(path: ":model:fn-execution", configuration: "shadow")
   testImplementation project(path: ":sdks:java:core", configuration: "testRuntimeMigration")
+  testImplementation project(path: ":sdks:java:extensions:avro", configuration: "testRuntimeMigration")
```
Comment (mosche) [inline runners/core-construction-java/build.gradle:77]: implementation should be enough

File: runners/google-cloud-dataflow-java/build.gradle [hunk 11]
```diff
@@ -115,6 +116,7 @@ dependencies {
   testImplementation library.java.junit
   testImplementation project(path: ":sdks:java:io:google-cloud-platform", configuration: "testRuntimeMigration")
   testImplementation project(path: ":sdks:java:core", configuration: "shadowTest")
+  testImplementation project(path: ":sdks:java:extensions:avro", configuration: "testRuntimeMigration")
```
Comment (mosche) [inline runners/google-cloud-dataflow-java/build.gradle:119]: just implementation?

File: sdks/java/io/parquet/build.gradle [hunk 12]
```diff
@@ -52,6 +53,7 @@ dependencies {
   provided library.java.hadoop_common
   testImplementation library.java.hadoop_client
   testImplementation project(path: ":sdks:java:core", configuration: "shadowTest")
+  testImplementation project(path: ":sdks:java:extensions:avro", configuration: "testRuntimeMigration")
```
Comment (mosche) [inline sdks/java/io/parquet/build.gradle:56]: just implementation?

**Review by mosche — APPROVED** [review]
Thanks, LGTM 🎉 
Though, still flagging the coder replacement of CountingSource as potential issue. I can't think of use cases where one would checkpoint this, but who knows ...

How do you want to proceed @aromanenko-dev?
Do you want to move ahead and merge or get more people to ack?

File: runners/spark/3/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/SparkSessionFactory.java (hunk 9 above)
Comment (aromanenko-dev) [reply runners/spark/3/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/SparkSessionFactory.java]: Makes sense, done

**Comment by aromanenko-dev** [comment]
@mosche Thanks, review comments are addressed.

> Though, still flagging the coder replacement of CountingSource as potential issue. I can't think of use cases where one would checkpoint this, but who knows ..

We can keep it "as it is" for now but it will be anyway a breaking change once Avro will be dropped from "core". Do you see any other options?

> Do you want to move ahead and merge or get more people to ack?

Thanks for review! Before merging, if possible, I'd be happy to have another look on this PR to make sure that we didn't miss something. 
Ping @kennknowles @reuvenlax @lukecwik

**Comment by mosche** [comment]
> We can keep it "as it is" for now but it will be anyway a breaking change once Avro will be dropped from "core". Do you see any other options?

@aromanenko-dev You don't need Avro to produce Avro compatible bytes. It's a well defined format with a detailed Spec. Of course this could become a hassle quickly, but luckily this case is trivial. Avro uses [varint zigzag encoding for longs](https://avro.apache.org/docs/1.8.2/spec.html#binary_encoding), that's just the same Protobuf is using for [signed ints](https://developers.google.com/protocol-buffers/docs/encoding?csw=1#signed-ints). And Avro records are just a concatenation of their fields without any further additions:

Here's an example using Protobufs `CodedOutputStream`
```Java
//import org.apache.beam.vendor.grpc.v1p48p1.com.google.protobuf.CodedOutputStream
CodedOutputStream cos = CodedOutputStream.newInstance(outputStream);
cos.writeSInt64NoTag(mark.getLastEmitted()); // signed int64 with varint zigzag encoding
cos.writeSInt64NoTag(mark.getStartTime().getMillis()); // signed int64 with varint zigzag encoding
cos.flush();
```

Of course, that would require some additional tests to verify compatibility...

**Comment by aromanenko-dev** [comment]
@mosche Imho, I think it's a overkill for this case but you have a point. So, I reverted `CountingSource.java` back and this issue may be addressed in #25252

**Comment by Abacn** [comment]
Hi @aromanenko-dev and @mosche, just wondering what is the status of this PR now? Because I also get a PR review close to merge (#24274) also made some changes on Avro.

**Comment by aromanenko-dev** [comment]
Run Java_GCP_IO_Direct PreCommit

**Comment by Abacn** [comment]
Java_GCP_IO_Direct failing due to `testTriggeredFileLoadsWithTempTablesToExistingNullSchemaTable[1]
` timeout flaky is known issue #25207

**Review by Abacn — APPROVED** [review]

**Comment by aromanenko-dev** [comment]
@Abacn Thanks for a ping. I think this PR is almost ready to be merged. In the mean time, the next step will be a deprecation of Avro classes in "sdks/java/core" but before, I think we need to sync them with "extensions/avro". WDYT?

PS: I also opened another PR #25216 to test "extensions/avro" against multiple Avro versions but it's WIP

**Comment by aromanenko-dev** [comment]
Run Java PostCommit

**Comment by Abacn** [comment]
> @Abacn Thanks for a ping. I think this PR is almost ready to be merged. In the mean time, the next step will be a deprecation of Avro classes in "sdks/java/core" but before, I think we need to sync them with "extensions/avro". WDYT?
> 

Yeah sounds good to me. As part of deprecation, I am thinking about if we can add a checkStyle guard to prevent new changes in beam repo referring to avro in core, like what we are doing for vendored dependencies (e.g. guava, grpc).

**Comment by aromanenko-dev** [comment]
@Abacn Interesting about checkStyle, I may try to do something similar for this. Thanks

**Comment by aromanenko-dev** [comment]
Run Java_Kafka_IO_Direct PreCommit

**Comment by Abacn** [comment]
Run Java_GCP_IO_Direct PreCommit

File: CHANGES.md [hunk 13]
```diff
@@ -60,6 +60,7 @@
   container was based upon Debian 11.
 * RunInference PTransform will accept model paths as SideInputs in Python SDK. ([#24042](https://github.com/apache/beam/issues/24042))
 * RunInference supports ONNX runtime in Python SDK ([#22972](https://github.com/apache/beam/issues/22972))
+* Java SDK modules migrated to use `:sdks:java:extensions:avro` ([#24748](https://github.com/apache/beam/issues/24748))
```
Comment (Abacn) [inline CHANGES.md:63]:
```suggestion
* Java SDK modules migrated to use `:sdks:java:extensions:avro` ([#24748](https://github.com/apache/beam/issues/24748))
```

**Comment by aromanenko-dev** [comment]
Run Java_GCP_IO_Direct PreCommit

**Comment by aromanenko-dev** [comment]
Run Portable_Python PreCommit

**Comment by aromanenko-dev** [comment]
Run SQL_Java11 PreCommit

**Comment by aromanenko-dev** [comment]
Run SQL_Java11 PreCommit

**Comment by aromanenko-dev** [comment]
Run SQL_Java17 PreCommit

**Comment by aromanenko-dev** [comment]
Run Java_SingleStore_IO_Direct PreCommit

**Comment by aromanenko-dev** [comment]
Run Whitespace PreCommit

**Comment by Abacn** [comment]
Java11/Java17 test failures are due to #23210, see investigation done in #25566

**Comment by aromanenko-dev** [comment]
@Abacn Do you think we have to wait for a fix before merging this one?

**Comment by Abacn** [comment]
Thanks @aromanenko-dev for pinning. I think we can proceed given that those failures are understood to be not related. I also opened #25566

**Comment by Abacn** [comment]
Run SQL_Java11 PreCommit

**Comment by Abacn** [comment]
Run SQL_Java17 PreCommit

**Comment by aromanenko-dev** [comment]
retest this please

**Comment by aromanenko-dev** [comment]
Run Java_GCP_IO_Direct PreCommit

**Comment by aromanenko-dev** [comment]
Run Java_Kafka_IO_Direct PreCommit

**Comment by aromanenko-dev** [comment]
Run Java_HCatalog_IO_Direct PreCommit

**Comment by aromanenko-dev** [comment]
I merge this one since failed checks are not related and all others are green.

**Pull request merged**